apps/invoices/src/modules/app-configuration/config-v3.ts

```diff
@@ -0,0 +1,137 @@
+import { z } from "zod";
+import { test } from "vitest";
+import { EncryptedMetadataManager, SettingsManager } from "@saleor/app-sdk/settings-manager";
+
+const AddressSchema = z.object({
+  companyName: z.string(),
+  cityArea: z.string(),
+  countryArea: z.string(),
+  streetAddress1: z.string(),
+  streetAddress2: z.string(),
+  postalCode: z.string(),
+  city: z.string(),
+  country: z.string(),
+});
+
+const ChannelSchema = z.object({
+  slug: z.string().min(1),
+});
+
+const AddressOverrideSchema = z.object({
+  address: AddressSchema,
+  channel: ChannelSchema,
+});
+
+const RootConfigSchema = z.object({
+  overrides: z.record(AddressOverrideSchema),
+});
+
+/**
+ * Root model that can serialize to json and parse from json.
+ *
+ * Uses Zod to parse and validate
+ *
+ * Adds domain methods on top
+ */
+export class ConfigModel {
+  /**
+   * Stores its own data as deep, single json, structured and validated by zod
+   */
+  private rootData: z.infer<typeof RootConfigSchema> = { overrides: {} };
+
+  constructor(initialConfig?: z.infer<typeof RootConfigSchema>) {
+    /**
+     * Sets its own initial state but also allows to inject - then validate
+     */
+    if (initialConfig) {
+      this.rootData = RootConfigSchema.parse(initialConfig);
+    }
+  }
+
+  /**
+   * Can statically parse itself, ensures parse/serialize work together
+   */
+  static parse(serialized: string) {
```

Comment:
This method is a bit confusing to me. When would we use it?

Reply:
Parse is the opposite to serialize. Just like JSON.stringify / JSON.parse

For plain objects JSON works, for classes we need to ensure that:
1. we serialize only things we need (no methods etc)
2. we parse it properly - because we need to instantiate class, call constructor

you use it when you load it from metadata. so
1. Fetch metadata (result is string)
2. run `Config.parse(stringifiedConfig)` -> resulting Config class
3. mutate config
4. stringify config with `config.stringify()`
5. save string to metadata manager